### ISSUE TITLE
Enable NGHTTP2

### ISSUE THREAD

**Issue by Jan-E** [winlibs/cURL#5]
For the next update: build nghttp2.lib and add ENABLE_NGHTTP2=yes to the build line.
https://github.com/curl/curl/pull/1141

This is relevant for PHP7+, but does not hurt in PHP 5.6.

**Comment by weltling** [comment]
@Jan-E thanks for the ping. What is currently the profit having nghttp in the stable branch? Not that we have any implementations in the core, some extensions maybe? Or is it picked up automatically, without any extra API having to be used?

Thanks.

**Comment by Jan-E** [comment]
@weltling It is automatically picked up by the curl extension, if libcurl is compiled with it. Steffen of Apachelounge regularly builds curl with a dependency on nghttp2.dll:
https://www.apachelounge.com/viewtopic.php?t=6784

See this topic on how this was started:
https://www.apachelounge.com/viewtopic.php?t=6740

In my PHP-builds I include a static nghttp2.lib, by using a slight modification of Makefile.msvc. 

See https://wiki.php.net/rfc/curl_http2_push as well.

**Comment by Jan-E** [comment]
See http://marc.info/?l=php-internals&m=144137527617547&w=2 as well for the initial implementation of the nghttp2 support.

**Comment by weltling** [comment]
Thanks for the explanations, @Jan-E, and for the work on the libcurl nmake script. So there are some features used which i've overslept, most because by the time they was added the http/2 support was not available. Now that it is possible, ofc it could be added. Probably for 7 only, as 5.6 is in security mode soon, so should keep it stable. Also waiting for the file:// issue in the latest libcurl, so no upgrade yet. Will keep nghttp2 in mind for the next possible opportunity.

Thanks.

**Comment by Jan-E** [comment]
Fine with me. I wasn't aware of any file:// issue, but i see now that you have reported it back to the curl dev's: https://github.com/curl/curl/issues/1187

In the mean time, here is my Makefile.msvc for creating a static nghhtp2.lib. It must be called by a GNU make (not nmake). I know you aren't a fan of that, but all the compiling actions are done by MSVC's own tools: cl.exe, link.exe, lib.exe and rc.exe. See the discussion about that in https://www.apachelounge.com/viewtopic.php?p=31517#31517 and following messages.

Maybe Steffen uses CMake nowadays. I am using this Makefile.msvc for GNU/cygwin make.exe:
```
#
# GNU Makefile for nghttp2 / MSVC.
#
# By G. Vanem <gvanem@yahoo.no> 2013
# Updated 3/2015 by Remo Eichenberger @remoe
# The MIT License apply.
#

#
# Choose your weapons:
#  Set 'USE_CYTHON=1' to build and install the 'nghttp2.pyd' Python extension.
#
THIS_MAKEFILE := $(lastword $(MAKEFILE_LIST))

USE_CYTHON := 0

_VERSION	:= $(shell grep AC_INIT ../configure.ac | cut -d'[' -f3 | sed -e 's/-DEV//g' -e 's/], //g')
_VERSION	:= $(subst ., ,$(_VERSION))
VER_MAJOR	:= $(word 1,$(_VERSION))
VER_MINOR	:= $(word 2,$(_VERSION))
VER_MICRO	:= $(word 3,$(_VERSION))
VERSION		:= $(VER_MAJOR).$(VER_MINOR).$(VER_MICRO)
VERSION_NUM := (($(VER_MAJOR) << 16) + ($(VER_MINOR) << 8) + $(VER_MICRO))

GENERATED   := 'Generated by $(realpath Makefile.MSVC)'

OBJ_DIR := MSVC_obj
#SUFFIX :=-vc90-mt-x86

#
# Where to copy nghttp2.dll + lib + headers to.
# Note: 'make install' is not in default targets. Do it explicitly.
#
TARGET_DIR  ?= ../_VC_ROOT
VC_ROOT     := $(abspath $(TARGET_DIR))
INSTALL_BIN := $(VC_ROOT)/bin
INSTALL_LIB := $(VC_ROOT)/lib
INSTALL_HDR := $(VC_ROOT)/include
DLL_R       := $(OBJ_DIR)/nghttp2$(SUFFIX).dll
DLL_D       := $(OBJ_DIR)/nghttp2d$(SUFFIX).dll
LIB_R       := $(OBJ_DIR)/nghttp2.lib
LIB_D       := $(OBJ_DIR)/nghttp2d.lib
IMP_R       := $(OBJ_DIR)/nghttp2-shared.lib
IMP_D       := $(OBJ_DIR)/nghttp2d-shared.lib

#
# Build for DEBUG-model and RELEASE at the same time.
#
TARGETS := $(LIB_R)

EXT_LIBS = 

NGHTTP2_PDB_R := $(OBJ_DIR)/nghttp2-shared.pdb
NGHTTP2_PDB_D := $(OBJ_DIR)/nghttp2d-shared.pdb

CC       = cl
LD       := link
AR       := lib
#CC      := icl
#LD      := xilink
#AR      := xilib
RC       := rc
CFLAGS   := -I./includes -Dssize_t=long -D_U_=""

CFLAGS_R := -nologo -MD  -W3 -Z7 -DBUILDING_NGHTTP2
CFLAGS_D := -nologo -MDd -W3 -Z7 -DBUILDING_NGHTTP2 \
           -Ot -D_DEBUG -GF -RTCs -RTCu # -RTCc -GS

LDFLAGS := -nologo -MAP -debug -incremental:no -opt:ref,icf -MANIFEST # -verbose


NGHTTP2_SRC := nghttp2_pq.c \
  nghttp2_map.c \
  nghttp2_queue.c \
  nghttp2_frame.c \
  nghttp2_buf.c \
  nghttp2_stream.c \
  nghttp2_outbound_item.c \
  nghttp2_session.c \
  nghttp2_submit.c \
  nghttp2_helper.c \
  nghttp2_npn.c \
  nghttp2_hd.c \
  nghttp2_hd_huffman.c \
  nghttp2_hd_huffman_data.c \
  nghttp2_version.c \
  nghttp2_priority_spec.c \
  nghttp2_option.c \
  nghttp2_callbacks.c \
  nghttp2_mem.c \
  nghttp2_http.c \
  nghttp2_rcbuf.c

NGHTTP2_OBJ_R := $(addprefix $(OBJ_DIR)/r_, $(notdir $(NGHTTP2_SRC:.c=.obj)))
NGHTTP2_OBJ_D := $(addprefix $(OBJ_DIR)/d_, $(notdir $(NGHTTP2_SRC:.c=.obj)))

.PHONY: all intro test_ver install copy_headers_and_libs \
	install_nghttp2_pyd_0 install_nghttp2_pyd_1 \
	build_nghttp2_pyd_0 build_nghttp2_pyd_1 \
	clean_nghttp2_pyd_0 clean_nghttp2_pyd_1


all: intro includes/nghttp2/nghttp2ver.h $(OBJ_DIR) $(TARGETS) build_nghttp2_pyd_$(USE_CYTHON)
	@echo 'Welcome to NgHTTP2 (release + debug).'
	@echo 'Do a "make -f Makefile.MSVC install" at own risk!'

intro:
	@echo 'Building NgHTTP (MSVC) ver. "$(VERSION)".'

test_ver:
	@echo '$$(VERSION):   "$(VERSION)".'
	@echo '$$(_VERSION):  "$(_VERSION)".'
	@echo '$$(VER_MAJOR): "$(VER_MAJOR)".'
	@echo '$$(VER_MINOR): "$(VER_MINOR)".'
	@echo '$$(VER_MICRO): "$(VER_MICRO)".'

$(OBJ_DIR):
	- mkdir $(OBJ_DIR)

install: $(TARGETS) \
         copy_headers_and_libs install_nghttp2_pyd_$(USE_CYTHON) 

#
# This MUST be done before using the 'install_nghttp2_pyd_1' rule.
#
copy_headers_and_libs:
	- mkdir -p $(INSTALL_HDR)/nghttp2 $(INSTALL_BIN) $(INSTALL_LIB)
	rm -f $(INSTALL_BIN)/nghttp2*
	rm -f $(INSTALL_LIB)/nghttp2*
	cp $(addprefix includes/nghttp2/, nghttp2.h nghttp2ver.h) $(INSTALL_HDR)/nghttp2
	cp $(LIB_R) $(INSTALL_LIB)
	@echo

$(LIB_R): $(NGHTTP2_OBJ_R)
	$(AR) -nologo -out:$@ $^
	@echo

$(LIB_D): $(NGHTTP2_OBJ_D)
	$(AR) -nologo -out:$@ $^
	@echo


$(IMP_R): $(DLL_R)

$(DLL_R): $(NGHTTP2_OBJ_R)/cygdrive/c/php-sdk/win 
	$(LD) $(LDFLAGS) -dll -out:$@ -implib:$(IMP_R) $(NGHTTP2_OBJ_R) -PDB:$(NGHTTP2_PDB_R)/cygdrive/c/php-sdk/win $(EXT_LIBS)
	mt -nologo -manifest $@.manifest -outputresource:$@\;2
	@echo

$(IMP_D): $(DLL_D)
	
$(DLL_D): $(NGHTTP2_OBJ_D)/cygdrive/c/php-sdk/win
	$(LD) $(LDFLAGS) -dll -out:$@ -implib:$(IMP_D) $(NGHTTP2_OBJ_D) -PDB:$(NGHTTP2_PDB_D)/cygdrive/c/php-sdk/win$(EXT_LIBS)
	mt -nologo -manifest $@.manifest -outputresource:$@\;2
	@echo


WIN_OBJDIR:=$(shell cygpath -w $(abspath $(OBJ_DIR)))
WIN_OBJDIR:=$(subst \,/,$(WIN_OBJDIR))

../python/setup.py: ../python/setup.py.in $(THIS_MAKEFILE)
	cd ../python ; \
	echo '# $(GENERATED). DO NOT EDIT.' > setup.py ; \
	sed -e 's/@top_srcdir@/../'   \
	    -e 's%@top_builddir@%$(WIN_OBJDIR)%' \
	    -e 's/@PACKAGE_VERSION@/$(VERSION)/' setup.py.in >> setup.py ;

build_nghttp2_pyd_0: ;

build_nghttp2_pyd_1: $(addprefix ../python/, setup.py nghttp2.pyx)
	cd ../python ; \
	python setup.py build_ext -i -f bdist_wininst

install_nghttp2_pyd_0: ;
		
install_nghttp2_pyd_1: $(addprefix ../python/, setup.py nghttp2.pyx)
	cd ../python ; \
	pip install .

clean_nghttp2_pyd_0: ;

clean_nghttp2_pyd_1:
	cd ../python ; \
	rm -fR build dist

$(OBJ_DIR)/r_%.obj: %.c $(THIS_MAKEFILE)
	$(CC) $(CFLAGS_R) $(CFLAGS) -Fo$@ -c $<
	@echo

$(OBJ_DIR)/d_%.obj: %.c $(THIS_MAKEFILE)
	$(CC) $(CFLAGS_D) $(CFLAGS) -Fo$@ -c $<
	@echo

$(OBJ_DIR)/r_nghttp2.res: $(OBJ_DIR)/nghttp2.rc $(THIS_MAKEFILE)
	$(RC) -D_RELEASE -Fo $@ $<
	@echo

$(OBJ_DIR)/d_nghttp2.res: $(OBJ_DIR)/nghttp2.rc $(THIS_MAKEFILE)
	$(RC) -D_DEBUG -Fo $@ $<
	@echo

includes/nghttp2/nghttp2ver.h: includes/nghttp2/nghttp2ver.h.in $(THIS_MAKEFILE)
	sed < includes/nghttp2/nghttp2ver.h.in     \
	     -e 's/@PACKAGE_VERSION@/$(VERSION)/g' \
	     -e 's/@PACKAGE_VERSION_NUM@/$(VERSION_NUM)\
\#define NGHTTP2_STATICLIB 1/g' > $@
	touch --reference=includes/nghttp2/nghttp2ver.h.in $@


define RES_FILE
  #include <winver.h>

  VS_VERSION_INFO VERSIONINFO
    FILEVERSION    $(VER_MAJOR), $(VER_MINOR), $(VER_MICRO), 0
    PRODUCTVERSION $(VER_MAJOR), $(VER_MINOR), $(VER_MICRO), 0
    FILEFLAGSMASK  0x3fL
    FILEOS         0x40004L
    FILETYPE       0x2L
    FILESUBTYPE    0x0L
  #ifdef _DEBUG
    #define        VER_STR  "$(VERSION).0 (MSVC debug)"
    #define        DBG      "d"
    FILEFLAGS      0x1L
  #else
    #define        VER_STR  "$(VERSION).0 (MSVC release)"
    #define        DBG      ""
    FILEFLAGS      0x0L
  #endif
  BEGIN
    BLOCK "StringFileInfo"
    BEGIN
      BLOCK "040904b0"
      BEGIN
        VALUE "CompanyName",      "http://tatsuhiro-t.github.io/nghttp2/"
        VALUE "FileDescription",  "nghttp2; HTTP/2 C library"
        VALUE "FileVersion",      VER_STR
        VALUE "InternalName",     "nghttp2" DBG
        VALUE "LegalCopyright",   "The MIT License"
        VALUE "LegalTrademarks",  ""
        VALUE "OriginalFilename", "nghttp2" DBG ".dll"
        VALUE "ProductName",      "NGHTTP2."
        VALUE "ProductVersion",   VER_STR
      END
    END
  BLOCK "VarFileInfo"
  BEGIN
    VALUE "Translation", 0x409, 1200
  END
  END
endef

export RES_FILE

$(OBJ_DIR)/nghttp2.rc: Makefile.MSVC
	@echo 'Generating $@...'
	@echo ' /* $(GENERATED). DO NOT EDIT.' > $@
	@echo '  */'                          >> $@
	@echo "$$RES_FILE"                    >> $@

clean:
	rm -f $(OBJ_DIR)/* includes/nghttp2/nghttp2ver.h
	@echo

vclean realclean: clean clean_nghttp2_pyd_$(USE_CYTHON)
	- rm -rf $(OBJ_DIR)
	- rm -f .depend.MSVC

#
# Use gcc to generated the dependencies. No MSVC specific args please!
#
REPLACE_R = 's/\(.*\)\.o: /\n$$(OBJ_DIR)\/r_\1.obj: /'
REPLACE_D = 's/\(.*\)\.o: /\n$$(OBJ_DIR)\/d_\1.obj: /'

depend: includes/nghttp2/nghttp2ver.h
	@echo '# $(GENERATED). DO NOT EDIT.' > .depend.MSVC
	gcc -MM $(CFLAGS) $(NGHTTP2_SRC)    >> .depend.tmp
	@echo '#'                           >> .depend.MSVC
	@echo '# Release lib objects:'      >> .depend.MSVC
	sed -e $(REPLACE_R) .depend.tmp     >> .depend.MSVC
	@echo '#'                           >> .depend.MSVC
	@echo '# Debug lib objects:'        >> .depend.MSVC
	sed -e $(REPLACE_D) .depend.tmp     >> .depend.MSVC
	rm -f .depend.tmp

-include .depend.MSVC
```

**Comment by weltling** [comment]
@Jan-E libcurl 7.53.0 with nghtp2 has now landed in the staging series. Please check.

Thanks.

**Comment by Jan-E** [comment]
Looks OK, as long as you distribute nghttp2.dll along with the PHP builds. All extensions that use libcurl_a.lib will depend on it. That is, for instance: php_curl.dll, php_oauth.dll, php_http.dll (pecl_http) and php_solr.dll. You might consider to build nghttp.lib statically into libcurl_a.lib, otherwise people will be surprised by a new dependency when they recompile php_http.dll with the new libcurl_a.lib.

BTW: tomorrow there will be a patch release: 7.53.1

**Comment by weltling** [comment]
Deps are always a complete package, or fetched completely if the new binary SDK is used. Thus new core builds are not an issue. But the new pecl builds with older PHP release - that could be a wtf, yep, once deps are updated on the build host. One more reason could be, that DLL from Apache would override PHP's ones. On the other hand, other DLLs like OpenSSL are overridden all the time. Got to experiment with this, probably patch cmake scripts. Holding on then till the next cURL is otu.

Thanks.

**Comment by Jan-E** [comment]
This is why I am often seduced to use static libs. The only real exception: the 2 OpenSSL dll's. But even these differ from the ones, used by PHP, because I compiled them with OpenSSL FIPS 2.0.14. And I always keep the DLL's in sync between Apache/bin and the TS PHP version (of the same VC).

BTW: Will PHP 5.6.30 on windows.php.net be recompiled with the new libcurl? After all, it is a security issue in one of the core extensions.

**Comment by weltling** [comment]
Ok, seems we pass on this upgrade. Currently libcurl 7.51.0 is used, which is not affected by the security issue. I'm leaving 7.53.0+nghttp2 for master staging only, might rework in between in accordance to what was discussed. Expect some upgrade later on in march.

Thanks for the checks.

**Comment by weltling** [comment]
@Jan-E in cURL makefile, ENABLE_NGHTTP2 sounds wrong. We depend on an external library, so it has to be WITH_NGHTTP2. Further more - link with static nghttp2 requires passing additional flags which is not done on the cURL side :( I see, that you pass the static flag in your makefile, but it is not going to work with the standard cURL build. I think, libcurl needs a correction, as the NGHTTP2 support is based on non official build.

Thanks.

**Comment by weltling** [comment]
@jan-e please check https://github.com/curl/curl/pull/1321.

Thanks.

**Comment by Jan-E** [comment]
The only, minor problem I have with https://github.com/curl/curl/pull/1321 is that I always build Curl and Apache with the static nghttp2 lib. I do not want to be dependent on a nghttp2.dll somewhere on my system.

Apache has 'nghttp2.lib' (i.e. the dll-version of nghttp2) hard coded in the Windows build files:
http://svn.apache.org/repos/asf/httpd/httpd/branches/2.4.x/modules/http2/

See for instance the LINK32_FLAGS in
http://svn.apache.org/repos/asf/httpd/httpd/branches/2.4.x/modules/http2/mod_http2.mak

After this PR I'll need nghttp2_static.lib for building Curl and the same file renamed to nghttp2.lib for building Apache. Minor problem, which can be solved easily in my Apache build scripts.

**Comment by weltling** [comment]
Thanks for checking. We'd better discuss this in the CURL PR. But anyway. It should not change anything for you, if accepted, as i left the enable flag and you trick out the compiler define on your own into the header. There's also another PR that targets to make lib names configurable https://github.com/curl/curl/pull/1259, if accepted.

In general, the approaches should target what the projects do by default, not the tricks that we do ourselves. As otherwise it creates even more mess in already messy Windows libs world :) Fe nghttp2.lib is created by default as an import library, not as a static lib. Getting a static library built needs also cunit, not sure why it's done so but it is the current nghttp2 state. I'd therefore stay by the shared nghttp2 for now, with somewhat patching to support the current PHP naming scheme. This way it's indeed less build patching need on both cURL, nghttp2 and PHP side. Smaller binary size is of a plus as well. As OpenSSL and possibly other libs are already being overridden in cases like Apache, so there should be no big issue if nghttp2 were. Anyway, probably keeping on this for now, especially at least as long as the other PR at cURL not (yet or if) accepted.

Thanks.

**Comment by Jan-E** [comment]
I saw now that in the QA releases 7.1.4RC1 has nghttp2.dll and 7.0.18RC1 does not. Do you plan to integrate nghttp2 only in the 7.1 releases?

**Comment by weltling** [comment]
Nope, should be in all 7.x branches. It's also easier, as otherwise we'd have to produce different libcurl builds of the same libcurl version.

The way to get it is a bit more fine grained, however. 7.0 doesn't require HTTP2 support in ext/curl, so cURL 7.53.1 is only for staging there. Already included in the AppVeyor runs, but omited for the release. If everything goes good, the next required libcurl upgrade should have nghttp2 in all the branches.

Thanks.

**Comment by Jan-E** [comment]
Next curl update is planned on April 19. Of course not yet clear if it will be required to update. Anyway, that is too late for 7.1.4 and 7.0.18..

**Comment by weltling** [comment]
Yep, no pressure with 7.0 anyway. Better we're stable than too fast.

Thanks.
